Added page to show how to use Vertical Pod Autoscaler recommendations (#135)
appuio/appuio-cloud-docs: docs/modules/ROOT/pages/how-to/use-vpa.adoc

= How to use Vertical Pod Autoscaler to adjust Pod Resouce levels

It's complicated to find out good values for xref:how-to/check-cpu-requests.adoc[investigate and resolve fair use ratio violations for a project]. APPUiO Cloud offers the OpenShift https://docs.openshift.com/container-platform/4.11/nodes/pods/nodes-pods-vertical-autoscaler.html[Vertical Pod Autoscaler (VPA)] object to assist DevOps engineers to fine tune the requests and limits of their deployments.

> This document explains how to use VPAs. This document assumes that you have a working project with some payload deployed and running.

Reviewer: ```suggestion
This document explains how to use VPAs. It assumes that you have a working project with some payload deployed and running.
```

== Limitations

The VPA operator is provided in APPUiO Cloud with the following limitations:

* The VPA operator only provides recommendations; it doesn't recreate pods automatically with new request and limit values.

== Create the VPA Object

Use the YAML below to define a new VPA object, and `oc apply` it to your namespace.

[source,yaml]
--
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: vpa-recommender
spec:
  targetRef:
    apiVersion: "apps/v1"
    kind:       Deployment
    name:       vpa-example  # <1>
  updatePolicy:

>     updateMode: "Auto"

Reviewer 2: I'd use `updateMode: Off` to ensure that existing configurations don't start recreating pods if we decide to deploy the `vpa-updater` as well at some point.

```suggestion
    updateMode: "Off"
```

[source,bash]
--
$ oc get vpa vpa-recommender --output yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  annotations: …
# …
spec:
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: vpa-example
  updatePolicy:
    updateMode: Auto
status:
  conditions:
  - status: "True"
    type: RecommendationProvided
  recommendation:
    containerRecommendations:
    - containerName: fortune-container
      lowerBound:
        cpu: 25m
        memory: 262144k
      target: # <1>
        cpu: 203m
        memory: 262144k
      uncappedTarget:
        cpu: 203m
        memory: 262144k
      upperBound: # <2>
        cpu: 71383m
        memory: "6813174422"
--
<1> Use this value as `request` in your deployment.
<2> Use this value as `limit` in your deployment.

== Interpreting the Recommendations

You should use the values provided by the autoscaler in your `Deployment` as follows:

* The `status.recommendation.containerRecommendations[*].target` value should be used as `request` values;

> * The `status.recommendation.containerRecommendations[*].upperBound` value should be used as a sensible `limit`.

Reviewer: Looking at this example, I shy back from this suggestion. Or at least, add a fat warning to have a closer look at the numbers and do not apply them blindly.

Maybe cross-reference with https://docs.appuio.cloud/user/how-to/check-cpu-requests.html and the dashboard mentions there. That dashboard shows utilization numbers for both CPU and memory limits. I see those numbers as a second source to find suitable limits.

I guess the discrepancy her is: VPA only touches requests but not limits. The uppreBound only guides VPA when to scale down requests and when not. So using it as a guide for limits might be a bit of a stretch.

> * The `status.recommendation.containerRecommendations[*].upperBound` value should be used as a sensible `limit`.

Author: Please check the latest commits and rephrasing

<2> The VPA is deployed to only support recomendation mode. So `Off` is the only supported value there.

== Find Recommendations

The VPA requires a few moments to gather data and provide recommendations from it. After some time, during which your deployment should have been running in order to gather meaningful data, run the following command and you should see output similar to this in your terminal:

[source,bash]
--
$ oc get vpa vpa-recommender --output yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  annotations: …
# …
spec:
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: vpa-example
  updatePolicy:
    updateMode: Auto
status:
  conditions:
  - status: "True"
    type: RecommendationProvided
  recommendation:
    containerRecommendations:
    - containerName: fortune-container
      lowerBound:
        cpu: 25m
        memory: 262144k
      target: # <1>
        cpu: 203m
        memory: 262144k
      uncappedTarget:
        cpu: 203m
        memory: 262144k
      upperBound: # <2>
        cpu: 71383m
        memory: "6813174422"
--
<1> Use this value as `request` in your deployment.

> <2> Use this value as `limit` in your deployment.

Reviewer 2: It may be worth it to repeat ourselves here

```suggestion
<2> Consider using this value as `limit` in your deployment.
Analyze the reported `upperBound` carefully before using it as the `limit` in your deployment.
The recommender will start with a very high upper bound, and update it over time as it observes the running application.
```
